fdbserver/VersionedBTree.actor.cpp

```diff
@@ -6151,42 +6170,46 @@ class VersionedBTree {
 							cursor.moveNext();
 						}
 					}
-				}
 
-				constexpr int maxHeightAllowed = 8;
-
-				// Write the new record(s) for the mutation boundary start key if its value has been set
-				// Clears of this key will have been processed above by not being erased from the updated page or
-				// excluded from the merge output
-				if (applyBoundaryChange && mBegin.mutation().boundarySet()) {
-					RedwoodRecordRef rec(mBegin.key(), mBegin.mutation().boundaryValue.get());
-					changesMade = true;
+					// If the boundary value is being set and we must insert it, add it to the page or the output set
+					if (shouldInsertBoundary) {
+						RedwoodRecordRef rec(mBegin.key(), mBegin.mutation().boundaryValue.get());
+						changesMade = true;
 
-					// If updating, add to the page, else add to the output set
-					if (updating) {
-						// Copy page for modification if not already copied
-						if (!pageCopy.isValid()) {
-							pageCopy = clonePageForUpdate(page);
-							btPage = (BTreePage*)pageCopy->begin();
-							cursor.tree = btPage->tree();
+						// If updating, first try to add the record to the page
+						if (updatingInPlace) {
+							copyForUpdate();
+							if (cursor.insert(rec, update->skipLen, maxHeightAllowed)) {
+								btPage->kvBytes += rec.kvBytes();
+								debug_printf("%s Inserted %s [mutation, boundary start]\n",
+								             context.c_str(),
+								             rec.toString().c_str());
+							} else {
+								debug_printf("%s Insert failed for %s [mutation, boundary start]\n",
+								             context.c_str(),
+								             rec.toString().c_str());
+								switchToLinearMerge();
+							}
 						}
 
-						if (cursor.insert(rec, update->skipLen, maxHeightAllowed)) {
-							btPage->kvBytes += rec.kvBytes();
+						// If not updating, add record to the output set
+						if (!updatingInPlace) {
```

Comment:
Why is it simply not the "else {"?

Reply:
Because the attempt to insert can fail due to lack of available space in the DeltaTree, in which case we must switch to rebuild mode.  `switchToLinearMerge()` executes this switch and populates the `merge` vector with all the data so far.  It also sets `updatingInPlace` to false.

Reply:
Ah, ok. Makes sense. Again, this is a very useful explanation to add as a comment :)

Reply:
I added a comment for this, but also `switchToLinearMerge()` used to be called from several places and now it is only one, so I'll just inline it which also makes it more clear.

Reply:
Perfect